vignettes/datatable-benchmarking.Rmd

```diff
@@ -17,11 +17,56 @@ h2 {
 }
 </style>
 
-This document is meant to guide on measuring performance of data.table. Single place to documents best practices or traps to avoid.
+This vignette documents the best ways to accurately measure data.table's performance, as well as common traps to avoid.
+
+## `microbenchmark` is not always appropriate
+Repetitive benchmarking like that used by `microbenchmark` may not always be appropriate. 
+While it makes perfect sense for timing atomic calculations, it may provide misleading results for common data processing tasks, which are typically larger than the scale for which `microbenchmark` was designed, and run only once.
+
+For example, the following gives the impression that `data.table` is an order of magnitude slower than `data.frame` for updating columns.
+
+``` r
+library(microbenchmark)
+library(data.table)
+DF <- data.frame(x = 1:5)
+DT <- data.table(x = 1:5)
+microbenchmark(DF$y <- seq.int(nrow(DF)), DT[, y := .I])
+#> Unit: microseconds
+#>                       expr     min      lq      mean   median       uq        max neval cld
+#>  DF$y <- seq.int(nrow(DF))   6.927   9.939  14.37229  15.2095  17.4690     35.840   100  a 
+#>          DT[, `:=`(y, .I)] 207.510 221.064 276.51600 228.7435 238.0795   4912.183   100   b
+```
+
+yet there is typically no real difference:
+
+```r
+n = 1e7
+DT = data.table( a=sample(1:1000,n,replace=TRUE),
+                 b=sample(1:1000,n,replace=TRUE),
+                 c=rnorm(n),
+                 d=sample(c("foo","bar","baz","qux","quux"),n,replace=TRUE),
+                 e=rnorm(n),
+                 f=sample(1:1000,n,replace=TRUE) )
+DF <- as.data.frame(DT)
+system.time(DF$y <- seq.int(nrow(DF)))
+#>    user  system elapsed 
+#>    0.01    0.00    0.02
+system.time(DT[, y := .I])
+#>    user  system elapsed 
+#>    0.01    0.01    0.03
+```
+
+Further, when benchmarking `set*` functions it only makes sense to measure the first run. Those functions update data.table by reference thus after the first run the data.table involved in the measurement will have changed.
```

Comment:
I understand this point from theoretical perspective, but is there a tangible evidence that setting something multiple times makes any performance difference? If yes, then this vignette is probably the best place to enunciate such examples. If no, then forcing a copy of data before each measurement creates more statistical noise than it purports to solve.